# Finding functional differences between species in a microbial community
This is a repository to accompany the publication <br />
**Abstract** <br />
Microbial life usually takes place in a community where individuals interact, by competition for nutrients, cross-feeding, inhibition by end-products, but also by their spatial distribution. Lactic acid bacteria are prominent members of microbial communities responsible for food fermentations. Their niche in a community depends on their own properties as well as those of the other species. Here, we apply a computational approach, which uses only genomic and metagenomic information and functional annotation of genes, to find properties that distinguish a species from others in the community, as well as to follow individual species in a community. We analyzed isolated and sequenced strains from a kefir community, and metagenomes from wine fermentations. We demonstrate how the distinguishing properties of an organism lead to experimentally testable hypotheses concerning the niche and the interactions with other species. We observe, for example, that L. kefiranofaciens, a dominant organism in kefir, stands out among the Lactobacilli because it potentially has more amino acid auxotrophies. Using metagenomic analysis of industrial wine fermentations we investigate the role of an inoculated L. plantarum in malolactic fermentation. We observed that L. plantarum thrives better on white than on red wine fermentations and has the largest number of phosphotransferase system among the bacteria observed in the wine communities. Also, L. plantarum together with Pantoea, Erwinia, Asaia, Gluconobacter, and Komagataeibacter genera had the highest number of genes involved in biosynthesis of amino acids.<br />
doi: https://doi.org/10.3389/fmicb.2019.01347

# Code
This folder contains code written in R and python for table generation and data analysis

# Data_accession
This folder contains accession number for the species from kefir community and accession number for metagenomics shotgun samples

# Post_processed_data
This folder contains most of the produces table generated for the analysis

# Publication_plots
This folder contains all the plots present in the publication

# Skin_experiment 
This folder contains information from the experiment performed with grape must and grape skins
